### PR TITLE
bpo-32990: Support WAVE_FORMAT_EXTENSIBLE in the wave module

### PR DESCRIPTION
**Pull request by ZackerySpytz**
The test file, a modified version of
Lib/test/audiodata/pluck-pcm24.wav, was provided by Andrea Celletti
on the bug tracker.

https://bugs.python.org/issue32990

### CODE REVIEW

File: Lib/wave.py [hunk 1]
```diff
@@ -254,19 +255,22 @@ def readframes(self, nframes):
...
-            except struct.error:
-                raise EOFError from None
-            self._sampwidth = (sampwidth + 7) // 8
-            if not self._sampwidth:
-                raise Error('bad sample width')
-        else:
+        if wFormatTag == WAVE_FORMAT_EXTENSIBLE:
+            # Only the first 2 bytes (of 16) of SubFormat are needed.
+            cbSize, wValidBitsPerSample, dwChannelMask, \
+            SubFormatFmt = struct.unpack_from('<HHLH', chunk.read(10))
```
Comment (zooba) [inline Lib/wave.py:266]: Could you reformat this line in a way that makes it easier to read? Right now it looks like two separate statements.

Using parentheses around the left hand side is fine, and then indent the second line.

Comment (zooba) [reply Lib/wave.py:266]: Oh, we should also catch `struct.error` here and raise, like we do above.

File: Lib/wave.py [hunk 2]
```diff
@@ -254,19 +255,22 @@ def readframes(self, nframes):
 
     def _read_fmt_chunk(self, chunk):
         try:
-            wFormatTag, self._nchannels, self._framerate, dwAvgBytesPerSec, wBlockAlign = struct.unpack_from('<HHLLH', chunk.read(14))
+            wFormatTag, self._nchannels, self._framerate, dwAvgBytesPerSec, \
+            wBlockAlign, wBitsPerSample = \
+                struct.unpack_from('<HHLLHH', chunk.read(16))
```
Comment (zooba) [inline Lib/wave.py:260]: Previously if there was no `wBitsPerSample` value and `wFormatTag` was something other than `WAVE_FORMAT_PCM`, we may get a different error.

Can we be sure that the longer `chunk.read` will always return 16 bytes? Or should we keep deferring the second read until after we've checked `wFormatTag`

**Comment by tiran** [comment]
FYI, we are planning to deprecate the wave module in 3.8 and move it out of the stdlib in 3.9 or 3.10. Does it make sense to add new features at all?

**Comment by ZackerySpytz** [comment]
@tiran As stated on the bug tracker, this could be considered a bug fix. It would be good to hear @zooba's opinion.

**Comment by jjuod** [comment]
Has there been a final decision on this? It seems the plan is to keep `wave` in, as of PEP 594, so maybe it's possible to finish this merge? We're using this through `wavio` to deal with various WAVs and could contribute a new PR, but this seems pretty close to merging already...

**Comment by csabella** [comment]
@ZackerySpytz, please address the code review.  Thank you!

**Comment by evenbrenden** [comment]
> Has there been a final decision on this? It seems the plan is to keep wave in, as of PEP 594, so maybe it's possible to finish this merge? We're using this through wavio to deal with various WAVs and could contribute a new PR, but this seems pretty close to merging already...

I'm in the same boat. Any objections to starting a new PR?

**Comment by ZackerySpytz** [comment]
I will update this PR.  A new one doesn't need to be created.

**Comment by zooba** [comment]
Same change was merged in #96777

**Pull request closed without merging**